### PR TITLE
Added Abbreviated Chat (/abc).

### PR DESCRIPTION
**Pull request by panserbjorne**
Condenses most user messages to improve readability.

Examples:
- "[S] [Soandso]: Test Message" (Say)
- "[O] [Soandso]: Test Message" (OOC)
- "[1] [Soandso]: Test Message" (Global channel 1)
- "<< [Soandso]: Test Message" (Tell from Soandso)
     - Still toying with other options such as [To] / [Fr]
- "[Soandso]: Test Message" (Tell when using tell windows)

Turned on/off with /abc 0-2. 
- 0 is disabled
- 1 is chat-only (to avoid breaking parsers)
- 2 is chat + logging

image

### CODE REVIEW

File: Zeal/chat.cpp [hunk 1]
```diff
@@ -633,6 +761,24 @@ Chat::Chat(ZealService *zeal) {
   //
   // return false;
   //}, callback_type::WorldMessage);
+  zeal->commands_hook->Add("/abbreviatedchat", {"/abc"}, "Abbreviates chat messages.",
```
Comment (CoastalRedwood) [inline Zeal/chat.cpp]: Can you add this command to the repo README.md?  The commands are alpha, so it would be at the top.

File: Zeal/tellwindows.cpp [hunk 2]
```diff
@@ -170,38 +218,55 @@ std::string GetName(std::string &data) {
...
+        message, word_regex,
+        "<a WndNotify=\"153," + name + "\">" + name + "</a>");
   }
 }
 
 void TellWindows::AddOutputText(Zeal::GameUI::ChatWnd *&wnd, std::string &msg, short channel) {
   if (!ZealService::get_instance()->tells->setting_enabled.get())  // just early out if tell windows are not enabled
     return;
 
-  replaceNameLinks(msg);
```
Comment (CoastalRedwood) [inline Zeal/tellwindows.cpp]: Huh, I don't know what this was doing outside of the if case below. No one ever complained, maybe it was just at the end of the callbacks list.

File: Zeal/tellwindows.cpp [hunk 3]
```diff
@@ -170,38 +218,55 @@ std::string GetName(std::string &data) {
...
+      replaceNameLinks(msg);
       if (!tell_window) {
         std::string WinName = TellWindowIdentifier + name;
         std::string ini_name = Zeal::Game::get_ui_ini_filename();
         int font_size = 3;
         if (ini_name.length()) {
           IO_ini ui_ini(ini_name);
           font_size = ui_ini.getValue<int>("ChatManager", "ChatWindow0_FontStyle");
-          font_size = std::clamp(font_size, 1, 6);
+          font_size = std::clamp(font_size, 0, 6);
```
Comment (CoastalRedwood) [inline Zeal/tellwindows.cpp]: I didn't do this clamping before so not sure why it was at 1.  I think 0 is fine, but pls keep an eye on it.

File: Zeal/chat.cpp [hunk 4]
```diff
@@ -144,24 +253,43 @@ static void __fastcall PrintChat(int t, int unused, char *data, short color_inde
 
   // Perform extra copies to protect unwary callers against the potential buffer size growth.
   char buffer[2048];  // Client maximum buffer size for print chat calls.
+  const auto &abbreviated_chat = ZealService::get_instance()->chat_hook->UseAbbreviatedChat;
   const auto &timestamp_style = ZealService::get_instance()->chat_hook->TimeStampsStyle;
+
+  std::string log_buffer = "";
```
Comment (CoastalRedwood) [inline Zeal/chat.cpp]: nit (push back if this look wrong): I prefer minimizing functional changes for the default path and simplifying if/else to remove now redundant code, so how about:

```
  std::string abbreviated_buffer = (abbreviated_chat.get() > 0) ? abbreviateChat(data) : "";
  const char *chat_buffer = (abbreviated_chat.get() > 0) ? abbreviated_buffer.c_str() : data;
  const char *log_buffer = (abbreviated_chat.get() == 2) ? abbreviated_buffer.c_str() : data;

  // Perform extra copies to protect unwary callers against the potential buffer size growth.
  char buffer[2048];  // Client maximum buffer size for print chat calls.
  const auto &timestamp_style = ZealService::get_instance()->chat_hook->TimeStampsStyle;
  bool log_is_different = timestamp_style.get() || (chat_buffer != log_buffer);
  if (timestamp_style.get()) {
    std::string timestamp_buffer = generateTimestampedString(chat_buffer, timestamp_style.get() == 1);
    strncpy_s(buffer, timestamp_buffer.c_str(), sizeof(buffer));
  } else {
    strncpy_s(buffer, chat_buffer, sizeof(buffer));
  }
  ZealService::get_instance()->hooks->hook_map["PrintChat"]->original(PrintChat)(t, unused, buffer, color_index,
                                                                                 add_log && !log_is_different);

  if (add_log && log_is_different && *Zeal::Game::is_logging_enabled) {
    strncpy_s(buffer, log_buffer, sizeof(buffer));
    Zeal::Game::GameInternal::DoPercentConvert(t, unused, buffer, 0);
    Zeal::Game::log(buffer);
  }
```

**Review by CoastalRedwood — COMMENTED** [review]
Sorry for the slow review.  Complex regexp makes my brain melt.  I didn't see anything wrong with them or any obvious crash causes.  I can merge after you comment / resolve some of my minor suggestions.

File: Zeal/chat.cpp [hunk 5]
```diff
@@ -65,6 +82,98 @@ std::string StripSpecialCharacters(const std::string &input) {
...
+  }
+
+  if (std::regex_search(original_message, match, roll_result_pattern)) {
+    if (playerRolling.length() == 0)  // Just in case
+      playerRolling = "?????";
+    std::string rollMin = match[1].str();
+    std::string rollMax = match[2].str();
+    std::string rollResult = match[3].str();
+    std::string newMessage = playerRolling + " rolls " + rollResult + " (" + rollMin + "-" + rollMax + ")";
+    playerRolling = std::string(); // Clear it for the next person
```
[the remark this answers is not in the document]
Comment (panserbjorne) [reply Zeal/chat.cpp]: Yeah, I figured one edge case might be if you zone in-between messages.

I haven't been able to test it but hopefully [these lines](https://github.com/CoastalRedwood/Zeal/pull/60/commits/c8bb6cffd729957057d54890d88c6f4eccbb067c#diff-cff273f7b9161420025a64e6155e9dcd1b85d4036bcd5fbff1f42dc4fb81a6edR163-R164) should handle it in some capacity.

```
    if (playerRolling.length() == 0)  // Just in case
      playerRolling = "?????";
```

File: Zeal/tellwindows.cpp (hunk 2 above)
Comment (panserbjorne) [reply Zeal/tellwindows.cpp]: Yeah not sure either, I moved it to avoid dealing with message links during the abbreviation stage

File: Zeal/chat.cpp (hunk 4 above)
Comment (panserbjorne) [reply Zeal/chat.cpp]: Yeah that looks a lot cleaner, updated the code.

File: Zeal/chat.cpp [hunk 6]
```diff
@@ -633,6 +761,24 @@ Chat::Chat(ZealService *zeal) {
...
+                               int abbreviated_option;
+                               if (Zeal::String::tryParse(args[1], &abbreviated_option)) {
+                                 abbreviated_option = std::clamp(abbreviated_option, 0, 2);
+                                 UseAbbreviatedChat.set(abbreviated_option);
+                               }
+                               
+                             } else {
+                               UseAbbreviatedChat.set(UseAbbreviatedChat.get() > 0 ? 0 : 1);
+                             }
+                             return true;  // return true to stop the game from processing any further on this command,
```
[the remark this answers is not in the document]
Comment (panserbjorne) [reply Zeal/chat.cpp]: Added

**Comment by panserbjorne** [comment]
Thanks for having a look. Should be ready to go.
